### DISCUSSION TITLE
Still running app after clicking home button or display turn off

### DISCUSSION THREAD

**Discussion by beppo-ivel** [appcelerator/titanium_mobile#13633] [Q&A] [answered]
Do I have to write a [service](https://titaniumsdk.com/guide/Titanium_SDK/Titanium_SDK_How-tos/Platform_API_Deep_Dives/Android_API_Deep_Dives/Android_Services.html) to collect e.g. GeoLocation after clicking home button or the display turn off. I checked the [tracker-app](https://github.com/FokkeZB/Tracker/tree/master/app) but could not figure out the mechanism, that the author implements to track GeoLocation after display turn off.

Or would it be enough to skip `$.destroy` when Window closed?

**Comment by m1ga** [comment, marked as answer]
Check https://titaniumsdk.com/api/titanium/android/service.html#background-location-service-example
That app is 8 years old and the Android eco system changed a lot! There are way more permissions you'll have to accept now in order to access location in the background. And you have to run a foreground service.
First think I would add is `exitOnClose: false` so it will put the app to background all the time and not close the app. 

Also check https://github.com/m1ga/titanium_background_app

**Reply by beppo-ivel** [reply]
I sorted out the possibilities by the given examples (simple, notification, geolocation) and updated your titanium background app. If this project still has relevance I can give a pull request and you can review the code. 

Further I have question in the [geoLocation](https://titaniumsdk.com/api/titanium/android/service.html#background-location-service-example) example:
1) Why they mix simple- and manual mode?
> Manual mode is used when the [Titanium.Geolocation.Android.manualMode](https://titaniumsdk.com/api/titanium/geolocation/android.html#manualmode) flag is set to true. In manual mode, the accuracy property is not used...

2) In my opinion I would implement the `getLocation` function as a service. Because the service is still running when display turn off or the app is in background. But maybe I think in a wrong way? In my procedure I just used simple mode (ti.playservices).

view:
```
<Alloy>
  <Tab title="Location">
	<Window onOpen="onOpen">
		<Button id="btn_start" top="50" title="start" onClick="onClickStart"/>
		<Button id="btn_stop" top="90" title="stop" onClick="onClickStop"/>
	</Window>
  </Tab>
</Alloy>
```
controller:
```
function onOpen(){
    Ti.Geolocation.requestLocationPermissions(Ti.Geolocation.AUTHORIZATION_ALWAYS, e => {
    if (e.success) {

    } else {
      alert('Could not obtain location permissions.');
    }
  });
}
Ti.Geolocation.accuracy = Ti.Geolocation.ACCURACY_HIGH;
const SECOND = Ti.Geolocation.accuracy == 1 ? 10 : 30;
Alloy.Collections.instance('location');
let intent = Titanium.Android.createServiceIntent( { url: 'locationService.js' } );
intent.putExtra('interval', 1000*SECOND);
let service = Titanium.Android.createService(intent);

function onClickStart(e) {
  service.start();
}

function onClickStop(e) {
  Ti.API.error(JSON.stringify(Alloy.Collections.location));
  service.stop();
  $.destroy();
}
```
lib/service:

```
Ti.Geolocation.getCurrentPosition(recordGeoPoint);

function recordGeoPoint(e){
  let point = Alloy.createModel('location', {longitude: e.coords.longitude, latitude: e.coords.latitude, altitude: e.coords.altitude, timestamp: Date.now()});
  Alloy.Collections.location.add(point);
  let date = new Date(point.get("timestamp"));
  let dateFormat = date.getHours() + ":" + date.getMinutes()+ ":" + date.getSeconds();
  Ti.API.error("long:" + point.get("longitude") + "    time:" + dateFormat);
}

```
model:
```

exports.definition = {
    config: {
        "columns": {
            "timestamp": "INTEGER",
            "altitude": "REAL",
            "latitude": "REAL",
            "longitude": "REAL",
            "id": "INTEGER PRIMARY KEY AUTOINCREMENT"
        },
        "adapter": {
            "type": "sql",
            "collection_name": "location",
            "idAttribute": "id"
        }
    },
    extendModel: function(Model) {
      _.extend(Model.prototype, {
        transform: function () {
          let model = this;
          let transformed = model.toJSON();

          return transformed;
        }
      });
      return Model;
    },
    extendCollection: function(Collection) {
      _.extend(Collection.prototype, { // Extend, override or implement Backbone.Collection
       
      });
      return Collection;
    }
}

```

**Reply by m1ga** [reply]
1. I think its just to show the usage or it could be left over from an older example. Doesn't hurt but it won't be used
2. Looks correct but I've never used it so you'll have to try it.

Thanks for the info about my demo app. It was just a test for myself, no need to extend it. Unless you have a working iOS example 😄

**Reply by beppo-ivel** [reply]
> ... so you'll have to try it.

I already did and it worked. Another thing that strikes me is `Ti.Android.requestPermissions([ 'android.permission.ACCESS_BACKGROUND_LOCATION' ]...`. This line does not guaranteed that location event listener is running when the app runs in background.

And I can't see the difference between implementing `createService` and a single thread like `setInterval`, where I can do the same logic. Is the interval cleared by the OS randomly compare to the service? 

>... Unless you have a working iOS example

Don't have a DE for iOS yet. AFAIK I need a MacOS for Xcode to deploy on iPhone emulator, right? Do you use virtual machine for this purpose?

**Reply by m1ga** [reply]
Virtual machine: checkout this post https://titanium.dzzd.net/t/developping-building-for-ios-from-windows-or-linux/64

**Reply by beppo-ivel** [reply]
> And you have to run a foreground service.

After deploying test app to mobile I recognized, that forgroundNotify service  is necessary in this [ example](https://titaniumsdk.com/api/titanium/android/service.html#background-location-service-example)  to execute service frequently if the app runs in background or display turns off. If forgroundNotify is missing the app runs also in background but not frequently as defined via service interval. It looks strage to implement a service that calls forgroundNotify to pull oneself up by one's bootstrap (to run correctly).
